### PR TITLE
fix: check full table name during logical plan creation

### PR DESCRIPTION
**Pull request by shuiyisong**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

1. move `Plugins` definition to `common-query` and pass `plugins` down to `QueryEngineState`
2. add `QueryOptions` to check whether full table name literal in sql should be strictly equal to user's catalog and schema in current `QueryContext`
  a. `Query` and `Explain` are be checked in `QueryEngineState` within datafusion's stmt to plan logic
  b. other SQLs are checked at frontend `query_statement`
3. add unit tests

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#909

### CODE REVIEW

File: src/frontend/src/instance.rs [hunk 1]
```diff
@@ -543,6 +570,20 @@ impl PromqlHandler for Instance {
     }
 }
 
+pub fn validate_catalog_and_schema(
```
Comment (sunng87) [inline src/frontend/src/instance.rs:573]: If we treat this as early form of our permission system, we can name this function as `check_permission` and accepts `&statement` and `query_ctx`

Comment (sunng87) [reply src/frontend/src/instance.rs:573]: And also `plugins`. Move `need_validate` into this function as well.

File: src/query/src/query_engine/options.rs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+use common_catalog::consts::DEFAULT_CATALOG_NAME;
+use datafusion_common::TableReference;
+use session::context::QueryContextRef;
+use snafu::ensure;
+
+use crate::error::{QueryAccessDeniedSnafu, Result};
+
+#[derive(Default)]
+pub struct QueryOptions {
+    pub validate_table_references: bool,
```
Comment (sunng87) [inline src/query/src/query_engine/options.rs:24]: we better make this option name more clear for its behaviour, and add some comments.

My suggestion: `disallow_cross_schema_query`

File: src/query/src/query_engine/options.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+pub fn validate_catalog_and_schema(
+    catalog: Option<&str>,
+    schema: &str,
+    query_ctx: &QueryContextRef,
+) -> Result<()> {
+    ensure!(
+        (catalog.is_none() || catalog.unwrap() == query_ctx.current_catalog())
```
Comment (sunng87) [inline src/query/src/query_engine/options.rs:33]: If catalog is none, do we need to check if current_catalog is the default catalog? If not, the catalog in error message below will be wrong.

File: src/common/query/src/lib.rs [hunk 4]
```diff
@@ -24,6 +24,8 @@ pub mod physical_plan;
 pub mod prelude;
 mod signature;
 
+pub type Plugins = anymap::Map<dyn core::any::Any + Send + Sync>;
```
Comment (sunng87) [inline src/common/query/src/lib.rs:27]: Consider move this to a more universal module liike `common/base`

Comment (Fengys123) [reply src/common/query/src/lib.rs:27]: +1

File: src/query/src/query_engine/options.rs (hunk 3 above)
Comment (shuiyisong) [reply src/query/src/query_engine/options.rs:33]: if catalog is none, should we put "" to err msg, e.g: "Illegal access to catalog: and schema: xxx"
or change `QueryAccessDenied` to receive a err msg and format it every time?

File: src/common/query/src/lib.rs (hunk 4 above)
Comment (MichaelScofield) [reply src/common/query/src/lib.rs:27]: How about creating a new crate "plugin"? It's wired to place "plugin"s either in "common/query" or "common/base". "base" is where we normally place some commonly used basic data structures and algorithms.

File: src/frontend/src/instance.rs [hunk 5]
```diff
@@ -543,16 +557,208 @@ impl PromqlHandler for Instance {
...
+    use std::collections::HashMap;
     use std::sync::atomic::AtomicU32;
 
     use session::context::QueryContext;
+    use strfmt::Format;
 
     use super::*;
     use crate::tests;
 
+    #[tokio::test(flavor = "multi_thread")]
```
Comment (Fengys123) [inline src/frontend/src/instance.rs:653]:
```suggestion
    #[tokio::test]
```

File: src/query/tests/query_engine_test.rs [hunk 6]
```diff
@@ -125,6 +124,39 @@ async fn test_udf() -> Result<()> {
     catalog_list
         .register_catalog(DEFAULT_CATALOG_NAME.to_string(), default_catalog)
         .unwrap();
+    Ok(catalog_list)
+}
+
+#[tokio::test]
```
Comment (Fengys123) [inline src/query/tests/query_engine_test.rs:130]:
```suggestion
#[test]
```

File: src/common/query/src/lib.rs (hunk 4 above)
Comment (sunng87) [reply src/common/query/src/lib.rs:27]: It's actually a basic data structure which be aliased to a scenario-connected name :joy: 

I'm afraid that a new crate to too heavy for an alias. Maybe put into `base` for now and move if the more operations are added?

File: src/query/src/query_engine/options.rs (hunk 3 above)
Comment (sunng87) [reply src/query/src/query_engine/options.rs:33]: We need to check how we deal with the situation when catalog is none:

- if it uses `DEFAULT_CATALOG_NAME` instead, check if query_ctx.current_catalog() == DEFAULT_CATALOG_NAME
- if it uses catalog from query_context, simply change error message to use query_ctx.current_catalog

Comment (shuiyisong) [reply src/query/src/query_engine/options.rs:33]: only `ShowTables` would call `validate_param` with `None` as catalog, and it actually use `query_ctx.current_catalog()` later on.

so I think the signature can be change to `fn validate_param(catalog: Option<&str>, schema: &str, query_ctx: &QueryContextRef)` and use `query_ctx.current_catalog()` directly

**Review by sunng87 — APPROVED** [review]

File: src/common/query/src/lib.rs (hunk 4 above)
Comment (shuiyisong) [reply src/common/query/src/lib.rs]: move definition to `common-base` for now. if more plugin-related structs are made, then we can create a new plugin crate 🤣

**Review by Fengys123 — APPROVED** [review]
LGTM

**Pull request merged**